sources/api/early-boot-config/src/bin/early-boot-config.rs

```diff
@@ -91,6 +95,85 @@ fn parse_args(args: env::Args) -> Args {
     }
 }
 
+/// Gather user data providers to run in order
+fn gather_providers() -> Result<Vec<PathBuf>> {
+    Ok(WalkDir::new(PROVIDERS_DIR)
+        .max_depth(1)
+        .min_depth(1)
+        .sort_by_file_name()
+        .into_iter()
+        .collect::<std::result::Result<Vec<_>, _>>()?
+        .into_iter()
+        .filter(|f| f.path().is_symlink())
+        .map(|f| f.into_path())
+        .collect())
+}
+
+/// Run a user data provider binary, forwarding its logs
+async fn run_provider<P>(log_level: &LevelFilter, provider: P) -> Result<process::Output>
+where
+    P: AsRef<Path>,
+{
+    let provider = provider.as_ref();
+    let result = AsyncCommand::new(provider)
+        .env(LOG_LEVEL_ENV_VAR, log_level.as_str())
+        .output()
+        .await
+        .context(error::CommandFailureSnafu {
+            provider: &provider,
+        })?;
+
+    // Regardless of provider status, log its output
+    let provider_name = provider
+        .file_name()
+        .unwrap_or(provider.as_os_str())
+        .to_string_lossy();
+    let provider_logs = String::from_utf8_lossy(&result.stderr);
+    for line in provider_logs.lines() {
+        info!("Provider '{}': {}", provider_name, line);
+    }
+
+    ensure!(
+        result.status.success(),
+        error::ProviderFailureSnafu {
+            provider: &provider,
+        }
```

Comment:
For error cases, the providers are emitting a string to stdout describing the error. We should probably include it here as part of the error message.